File: sdk/python/src/openagentio/bus/stream.py
"""Stream framing — server-side StreamWriter and client-side Stream iterator.

Mirrors pkg/bus/stream.go. A streaming response is a sequence of envelopes
sharing the same ``correlation_id``; the runtime tags each frame with a
monotonic ``seq`` and a terminal frame (``ResponseFinal`` or ``ResponseError``)
sets ``is_final = True``.
"""
from __future__ import annotations

import asyncio
import enum
from typing import Any

from openagentio.codec.json_codec import Codec
from openagentio.event.envelope import Envelope
from openagentio.bus.errors import (
    BackpressureDropError,
    error_code_for,
    is_retryable_for,
)
from openagentio.event.payload import ErrorPayload
from openagentio.event.types import (
    ResponseDelta,
    ResponseError,
    ResponseFinal,
    ResponseStarted,
)
from openagentio.transport.base import Inbox, RawMessage, Transport


# Defaults for the client-side out-of-order buffer. Mirror the Go SDK.
DEFAULT_MAX_PENDING_FRAMES = 256
DEFAULT_MAX_SEQUENCE_GAP = 1024


class ErrIdleTimeout(Exception):
    """Raised when the gap between two streaming frames exceeds the idle timeout."""


class _WriterState(enum.Enum):
    """Lifecycle of :class:`StreamWriter`.

    Mirrors ``writerState`` in ``pkg/bus/stream.go``. Prior to P0#3 the writer
    kept a single ``_closed`` boolean that was flipped before encode+publish;
    a codec/transport failure would then wedge the writer with no terminal
    frame on the wire. Modelling ``closing`` and ``failed`` as first-class
    states lets the runtime react (fallback error, structured logging) instead
    of silently swallowing the failure.
    """

    OPEN = "open"
    CLOSING = "closing"
    CLOSED = "closed"
    FAILED = "failed"


class StreamWriter:
    """Server-side stream emitter.

    Each method publishes one frame back to the request's ``reply_to`` subject.
    ``started`` may be called at most once. ``final`` and ``error`` are
    terminal — after either succeeds the writer transitions to ``CLOSED`` and
    subsequent calls raise. If encoding or publishing a terminal frame fails
    the writer transitions to ``FAILED`` instead: :meth:`final` / :meth:`error`
    raise the underlying exception, and the runtime may retry via
    :meth:`error` (permitted from both ``OPEN`` and ``FAILED``) to still land
    a terminal frame on the wire.
    """

    def __init__(
        self,
        transport: Transport,
        codec: Codec,
        agent_id: str,
        request: Envelope,
    ) -> None:
        self._transport = transport
        self._codec = codec
        self._agent_id = agent_id
        self._req = request
        self._lock = asyncio.Lock()
        self._seq = 0
        self._started = False
        self._state = _WriterState.OPEN
        # ``_last_err`` captures the most recent codec/publish failure so
        # bus._run_stream_handler can surface it when handler returned cleanly
        # but the terminal publish did not.
        self._last_err: BaseException | None = None
        # ``_reserved_seq`` holds the seq that a failed terminal frame
        # consumed. A fallback ``error(...)`` from ``FAILED`` must reuse that
        # seq so the client's reorder buffer sees a contiguous sequence —
        # otherwise the previously-reserved-but-never-published seq stays
        # missing forever and the client idle-times-out despite receiving
        # this terminal frame. ``_reserved_valid`` distinguishes "no
        # reservation" from "reservation is 0".
        self._reserved_seq: int = 0
        self._reserved_valid = False

    # --- state introspection ---------------------------------------------

    @property
    def state(self) -> _WriterState:
        return self._state

    @property
    def closed(self) -> bool:
        """Backwards-compat: returns True for any terminal state (closed OR
        failed). Callers that need to distinguish should check :attr:`state`.
        """
        return self._state in (_WriterState.CLOSED, _WriterState.FAILED)

    @property
    def last_error(self) -> BaseException | None:
        """Last codec/publish failure; ``None`` unless state is FAILED."""
        return self._last_err

    # --- terminal frames -------------------------------------------------

    async def started(self, meta: Any = None) -> None:
        async with self._lock:
            self._require_open_locked()
            if self._started:
                raise RuntimeError("stream: started already emitted")
            self._started = True
            seq = self._next_seq_locked()

        env = new_reply_shell(self._agent_id, self._req, ResponseStarted)
        env.seq = seq
        env.payload = self._codec.encode_payload(meta)
        await self._publish(env)

    async def delta(self, chunk: Any = None) -> None:
        async with self._lock:
            self._require_open_locked()
            seq = self._next_seq_locked()

        env = new_reply_shell(self._agent_id, self._req, ResponseDelta)
        env.seq = seq
        env.payload = self._codec.encode_payload(chunk)
        await self._publish(env)

    async def final(self, result: Any = None) -> None:
        """Publish a terminal ResponseFinal frame.

        Unlike the pre-P0#3 implementation, ``_closed`` is not set until
        after publish succeeds. If encoding or publishing fails the writer
        transitions to :attr:`_WriterState.FAILED`, the exception propagates,
        and the runtime is free to attempt a fallback :meth:`error`.
        """
        async with self._lock:
            self._require_open_locked()
            self._state = _WriterState.CLOSING
            seq = self._next_seq_locked()

        env = new_reply_shell(self._agent_id, self._req, ResponseFinal)
        env.seq = seq
        env.is_final = True
        try:
            env.payload = self._codec.encode_payload(result)
        except BaseException as e:
            await self._mark_failed(e, reserved_seq=seq)
            raise
        try:
            await self._publish(env)
        except BaseException as e:
            await self._mark_failed(e, reserved_seq=seq)
            raise
        await self._mark_closed()

    async def error(self, exc: BaseException) -> None:
        """Publish a terminal ResponseError frame.

        Permitted from ``OPEN`` (handler chose to fail) and ``FAILED``
        (previous Final could not reach the wire — try an error frame
        instead so the peer still sees a terminal). A publish failure here
        leaves the writer in ``FAILED`` and re-raises.
        """
        if exc is None:  # type: ignore[unreachable]
            raise ValueError("stream: error requires a non-None exception")

        async with self._lock:
            if self._state not in (_WriterState.OPEN, _WriterState.FAILED):
                raise self._terminal_state_error_locked()
            # Reuse the reserved seq from a prior failed terminal — see
            # ``__init__`` for the rationale. Without this, the fallback
            # error frame lands with the *next* seq while the previously
            # reserved seq stays missing on the wire; the client's reorder
            # buffer waits forever for a frame that will never arrive.
            if self._state is _WriterState.FAILED and self._reserved_valid:
                seq = self._reserved_seq
                self._reserved_valid = False
            else:
                seq = self._next_seq_locked()
            self._state = _WriterState.CLOSING

        env = new_reply_shell(self._agent_id, self._req, ResponseError)
        env.seq = seq
        env.is_final = True
        code = error_code_for(exc)
        retryable = is_retryable_for(exc)
        payload = ErrorPayload(code=code, message=str(exc), retryable=retryable)
        try:
            env.payload = self._codec.encode_payload(payload)
        except BaseException as e:
            await self._mark_failed(e, reserved_seq=seq)
            raise
        try:
            await self._publish(env)
        except BaseException as e:
            await self._mark_failed(e, reserved_seq=seq)
            raise
        await self._mark_closed()

    # --- internals -------------------------------------------------------

    def _require_open_locked(self) -> None:
        if self._state is _WriterState.OPEN:
            return
        raise self._terminal_state_error_locked()

    def _terminal_state_error_locked(self) -> RuntimeError:
        if self._state is _WriterState.CLOSING:
            return RuntimeError("stream: terminal frame in flight")
        if self._state is _WriterState.CLOSED:
            return RuntimeError("stream: already closed")
        if self._state is _WriterState.FAILED:
            return RuntimeError("stream: writer failed; only error is permitted")
        return RuntimeError(f"stream: unexpected state {self._state!r}")

    async def _mark_closed(self) -> None:
        async with self._lock:
            self._state = _WriterState.CLOSED
            self._last_err = None
            # A successful terminal consumed its seq; clear any prior
            # reservation defensively.
            self._reserved_valid = False

    async def _mark_failed(
        self, err: BaseException, *, reserved_seq: int | None = None
    ) -> None:
        async with self._lock:
            self._state = _WriterState.FAILED
            self._last_err = err
            if reserved_seq is not None:
                self._reserved_seq = reserved_seq
                self._reserved_valid = True
            else:
                self._reserved_valid = False

    def _next_seq_locked(self) -> int:
        s = self._seq
        self._seq += 1
        return s

    async def _publish(self, env: Envelope) -> None:
        data = self._codec.encode_envelope(env)
        await self._transport.publish(
            RawMessage(subject=self._req.reply_to, data=data)
        )


class Stream:
    """Async iterator over a streaming response.

    Frames are reordered by :py:attr:`Envelope.seq`; the iterator stops after
    yielding a frame with ``is_final = True``.

    Two timeouts coexist (mirroring the Go SDK):

    * ``idle_timeout`` — maximum gap between two frames; expiring raises
      :class:`ErrIdleTimeout`.
    * ``deadline`` — absolute wall-clock deadline for the whole stream
      (in :py:meth:`asyncio.AbstractEventLoop.time` units); expiring raises
      :class:`asyncio.TimeoutError`. The overall deadline always wins when
      both timers would fire — the iterator checks the deadline before and
      after each ``recv`` call.
    """

    def __init__(
        self,
        inbox: Inbox,
        codec: Codec,
        idle_timeout: float | None = None,
        deadline: float | None = None,
        max_pending_frames: int | None = None,
        max_sequence_gap: int | None = None,
    ) -> None:
        self._inbox = inbox
        self._codec = codec
        self._idle = idle_timeout
        self._deadline = deadline
        self._max_pending = (
            max_pending_frames
            if max_pending_frames is not None and max_pending_frames > 0
            else DEFAULT_MAX_PENDING_FRAMES
        )
        self._max_gap = (
            max_sequence_gap
            if max_sequence_gap is not None and max_sequence_gap > 0
            else DEFAULT_MAX_SEQUENCE_GAP
        )
        self._expected = 0
        self._pending: dict[int, Envelope] = {}
        self._exhausted = False

    def __aiter__(self) -> "Stream":
        return self

    async def __anext__(self) -> Envelope:
        if self._exhausted:
            raise StopAsyncIteration

        while True:
            ready = self._pending.pop(self._expected, None)
            if ready is not None:
                self._expected += 1
                if ready.is_final:
                    self._exhausted = True
                return ready

            wait = self._idle
            if self._deadline is not None:
                remaining = self._deadline - asyncio.get_running_loop().time()
                if remaining <= 0:
                    self._exhausted = True
                    raise asyncio.TimeoutError("bus: stream overall timeout")
                wait = remaining if wait is None else min(wait, remaining)

            try:
                msg = await self._inbox.recv(timeout=wait)
            except asyncio.TimeoutError:
                self._exhausted = True
                if (
                    self._deadline is not None
                    and asyncio.get_running_loop().time() >= self._deadline
                ):
                    raise asyncio.TimeoutError(
                        "bus: stream overall timeout"
                    ) from None
                raise ErrIdleTimeout("bus: stream idle timeout") from None
            except RuntimeError:
                # inbox closed mid-iteration — clean exit.
                self._exhausted = True
                raise StopAsyncIteration from None

            env = self._codec.decode_envelope(msg.data)
            # Reject negative Seq outright. Unlike the Go SDK (uint64), Python
            # envelopes carry Seq as a plain int and would otherwise treat a
            # negative value as a "late" frame that silently gets dropped;
            # that's a protocol violation, not a routing hiccup.
            if env.seq < 0:
                self._exhausted = True
                raise BackpressureDropError(
                    f"bus: negative seq={env.seq}"
                )
            if env.seq < self._expected:
                continue  # late / duplicate frame
            if env.seq in self._pending:
                continue
            # Backpressure guards — see pkg/bus/stream.go for rationale.
            #   1. Seq jumps too far ahead of expected: bound pending growth
            #      when a low-Seq frame is permanently missing.
            #   2. Pending is at capacity and this frame is not the expected
            #      one: accepting it would exceed the buffer. If it *is* the
            #      expected Seq, we take it — flushing it drains the buffer.
            #
            # Gap check uses subtraction rather than ``expected + max_gap`` for
            # symmetry with the Go SDK, where the same expression avoids a
            # uint64 overflow when ``expected`` is near ``MaxUint64``. Python's
            # ints are arbitrary-precision so overflow is not a concern here,
            # but keeping the two implementations line-for-line identical makes
            # them easier to audit together.
            if env.seq - self._expected >= self._max_gap:
                self._exhausted = True
                raise BackpressureDropError(
                    f"bus: stream seq gap too large (seq={env.seq}, "
                    f"expected={self._expected}, gap={self._max_gap})"
                )
            if env.seq != self._expected and len(self._pending) >= self._max_pending:
                self._exhausted = True
                raise BackpressureDropError(
                    f"bus: stream pending buffer full "
                    f"(max={self._max_pending}, seq={env.seq})"
                )
            self._pending[env.seq] = env

    async def close(self) -> None:
        if self._exhausted:
            return
        self._exhausted = True
        try:
            await self._inbox.close()
        except Exception:
            pass


def _inherit_metadata(src: dict[str, Any] | None) -> dict[str, Any] | None:
    if src is None:
        return None
    dst = {k: v for k, v in src.items() if not k.startswith("acp.")}
    return dst if dst else None


def new_reply_shell(agent_id: str, req: Envelope, event_type: str) -> Envelope:
    """Pre-populate a response envelope with correlation metadata copied from req.

    Non-acp metadata keys are inherited so business context (e.g. dingtalk.*)
    flows back through cascading invocations without manual copying.
    """
    resp = Envelope.new(event_type)
    resp.from_ = agent_id
    resp.to = req.from_
    resp.session_id = req.session_id
    resp.conversation_id = req.conversation_id
    resp.tenant_id = req.tenant_id
    resp.user_id = req.user_id
    resp.channel = req.channel
    resp.trace_id = req.trace_id
    resp.span_id = req.span_id
    resp.traceparent = req.traceparent
    resp.correlation_id = req.event_id
    resp.metadata = _inherit_metadata(req.metadata)
    return resp

File: pkg/bus/stream.go
package bus

import (
	"context"
	"errors"
	"iter"
	"sync"
	"time"

	"github.com/ModulationAI/openagentio/pkg/codec"
	"github.com/ModulationAI/openagentio/pkg/event"
	"github.com/ModulationAI/openagentio/pkg/middleware"
	"github.com/ModulationAI/openagentio/pkg/transport"
)

// StreamInvoke publishes a request to {prefix}.invoke.{target} with a fresh
// _INBOX as reply_to and returns a Stream over the received frames. Frames
// are reordered by Envelope.Seq; the iterator terminates after a frame with
// IsFinal=true, or earlier on idle/overall timeout.
func (b *defaultBus) StreamInvoke(ctx context.Context, target string, payload any, opts ...InvokeOption) (Stream, error) {
	if target == "" {
		return nil, errors.New("bus: empty invoke target")
	}
	o := collectInvokeOpts(opts)
	timeout := o.Timeout
	if timeout == 0 {
		timeout = b.opts.DefaultTimeout
	}

	streamCtx, cancel := context.WithCancel(ctx)
	if timeout > 0 {
		streamCtx, cancel = context.WithTimeout(ctx, timeout)
	}

	env, err := b.buildRequestEnvelope(target, payload)
	if err != nil {
		cancel()
		return nil, err
	}

	inbox, err := b.opts.Transport.OpenInbox(streamCtx)
	if err != nil {
		cancel()
		return nil, err
	}
	env.ReplyTo = inbox.Subject()

	b.prepareEnvelope(streamCtx, env)
	data, err := b.opts.Codec.EncodeEnvelope(env)
	if err != nil {
		_ = inbox.Close()
		cancel()
		return nil, err
	}
	if err := b.opts.Transport.Publish(streamCtx, &transport.RawMessage{
		Subject: b.invokeSubject(target, b.resolveTenant(env.TenantID)),
		Data:    data,
	}); err != nil {
		_ = inbox.Close()
		cancel()
		return nil, err
	}

	return &stream{
		ctx:        streamCtx,
		cancel:     cancel,
		inbox:      inbox,
		codec:      b.opts.Codec,
		idle:       o.IdleTimeout,
		maxPending: resolveMaxPending(o.MaxPendingFrames),
		maxGap:     resolveMaxGap(o.MaxSequenceGap),
	}, nil
}

func resolveMaxPending(n int) int {
	if n <= 0 {
		return DefaultMaxPendingFrames
	}
	return n
}

func resolveMaxGap(g uint64) uint64 {
	if g == 0 {
		return DefaultMaxSequenceGap
	}
	return g
}

// HandleStream subscribes to {prefix}.invoke.{target} and dispatches each
// request into a goroutine running the supplied handler. A StreamWriter is
// provided that publishes started/delta/final/error frames back to
// req.ReplyTo with monotonically increasing Seq numbers. If the handler
// returns without calling Final or Error, the runtime auto-emits one based on
// the returned error.
func (b *defaultBus) HandleStream(target string, h StreamHandler, opts ...HandleOption) error {
	if target == "" {
		return errors.New("bus: empty invoke target")
	}
	if h == nil {
		return errors.New("bus: nil stream handler")
	}
	o := collectHandleOpts(opts)
	if !o.QueueSet {
		o.Queue = target
	}
	subject := b.invokeSubject(target, b.opts.Tenant)

	dispatch := func(_ context.Context, msg *transport.RawMessage) error {
		req, err := b.opts.Codec.DecodeEnvelope(msg.Data)
		if err != nil {
			return err
		}
		if req.ReplyTo == "" {
			return errors.New("bus: stream request missing reply_to")
		}
		go b.handleStream(req, h)
		return nil
	}

	sub, err := b.opts.Transport.Subscribe(b.lifeCtx, subject, o.Queue, dispatch)
	if err != nil {
		return err
	}
	b.trackOwned(sub)
	return nil
}

func (b *defaultBus) handleStream(req *event.Envelope, h StreamHandler) {
	ctx, cancel := context.WithCancel(b.lifeCtx)
	defer cancel()

	w := &streamWriter{
		bus: b,
		ctx: ctx,
		req: req,
	}

	chained := middleware.Chain(middleware.Handler(func(c context.Context, e *event.Envelope) error {
		return h(c, e, w)
	}), b.opts.Middleware...)

	herr := chained(ctx, req)

	// The auto-terminal logic must handle every writer state — not just
	// `open`. If the handler called Final/Error and it succeeded, the writer
	// is `closed` and there is nothing to do. If Final/Error was attempted
	// but publish failed (writer is `failed`), the handler already saw the
	// error but the peer is still waiting on a terminal frame — we must try
	// once more with a synthetic Error so the client doesn't idle out.
	switch w.stateSnapshot() {
	case writerClosed:
		return
	case writerFailed:
		// Handler's Final/Error left no terminal frame on the wire.
		// Attempt a fallback Error so the peer transitions to a terminal
		// state instead of blocking on the idle timeout. Use handler's
		// error if any, else surface the last publish failure explicitly.
		fallback := herr
		if fallback == nil {
			fallback = w.lastErr()
		}
		if fallback == nil {
			fallback = errors.New("stream: terminal publish failed")
		}
		if ferr := w.forceError(fallback); ferr != nil {
			b.opts.Logger.Error("bus: stream fallback error publish failed",
				"target", subjectTargetFrom(req),
				"correlation_id", req.EventID,
				"reply_to", req.ReplyTo,
				"handler_err", herrString(herr),
				"publish_err", ferr,
			)
		}
	default: // writerOpen — handler returned without terminating
		var termErr error
		if herr != nil {
			termErr = w.Error(herr)
		} else {
			termErr = w.Final(nil)
		}
		if termErr != nil {
			b.opts.Logger.Error("bus: stream auto-terminal failed",
				"target", subjectTargetFrom(req),
				"correlation_id", req.EventID,
				"reply_to", req.ReplyTo,
				"handler_err", herrString(herr),
				"publish_err", termErr,
			)
		}
	}
}

// herrString renders a possibly-nil handler error for structured logging.
func herrString(e error) string {
	if e == nil {
		return ""
	}
	return e.Error()
}

// subjectTargetFrom extracts the invoke target for logging. Falls back to the
// event type when the target is not available on the request envelope.
func subjectTargetFrom(req *event.Envelope) string {
	if req == nil {
		return ""
	}
	if req.To != "" {
		return req.To
	}
	return req.EventType
}

// --- client-side stream ------------------------------------------------------

type stream struct {
	ctx        context.Context
	cancel     context.CancelFunc
	inbox      transport.Inbox
	codec      codec.Codec
	idle       time.Duration
	maxPending int
	maxGap     uint64

	// expected is the next Seq the iterator will yield. Promoted to a field
	// (rather than a local in Events()) so white-box tests can start the
	// iterator near uint64 boundaries without having to feed 2^64 frames.
	expected uint64

	closeOnce sync.Once
	closeErr  error
}

func (s *stream) Close() error {
	s.closeOnce.Do(func() {
		s.cancel()
		s.closeErr = s.inbox.Close()
	})
	return s.closeErr
}

func (s *stream) Events() iter.Seq2[*event.Envelope, error] {
	return func(yield func(*event.Envelope, error) bool) {
		pending := make(map[uint64]*event.Envelope)

		for {
			recvCtx := s.ctx
			var recvCancel context.CancelFunc
			if s.idle > 0 {
				recvCtx, recvCancel = context.WithTimeout(s.ctx, s.idle)
			}
			msg, err := s.inbox.Recv(recvCtx)
			if recvCancel != nil {
				recvCancel()
			}

			if err != nil {
				switch {
				case s.ctx.Err() != nil:
					yield(nil, s.ctx.Err())
				case s.idle > 0 && errors.Is(err, context.DeadlineExceeded):
					yield(nil, ErrIdleTimeout)
				default:
					yield(nil, err)
				}
				return
			}

			env, err := s.codec.DecodeEnvelope(msg.Data)
			if err != nil {
				yield(nil, err)
				return
			}

			if env.Seq < s.expected {
				continue // duplicate / late frame, drop
			}
			if _, dup := pending[env.Seq]; dup {
				continue
			}
			// Backpressure guards. Two conditions terminate the stream:
			//   1. Seq jumps too far ahead of the currently-expected one.
			//      This bounds unbounded pending accumulation when a low-Seq
			//      frame is permanently missing (or a malicious/buggy server
			//      injects a huge Seq).
			//   2. Pending buffer already at capacity and this frame is not
			//      the currently-expected Seq. If it *is* the expected one,
			//      we accept it — flushing it drains the buffer immediately.
			//
			// Gap check uses subtraction rather than `expected + maxGap` to
			// avoid uint64 overflow when expected is near MaxUint64: the
			// addition would wrap around and misclassify a valid frame as
			// over-gap. The `env.Seq < expected` branch above guarantees
			// `env.Seq - expected` never underflows here.
			if env.Seq-s.expected >= s.maxGap {
				yield(nil, ErrBackpressureDrop)
				return
			}
			if env.Seq != s.expected && len(pending) >= s.maxPending {
				yield(nil, ErrBackpressureDrop)
				return
			}
			pending[env.Seq] = env

			for {
				e, ok := pending[s.expected]
				if !ok {
					break
				}
				delete(pending, s.expected)
				s.expected++
				if !yield(e, nil) {
					return
				}
				if e.IsFinal {
					return
				}
			}
		}
	}
}

// --- server-side writer ------------------------------------------------------

// writerState models the lifecycle of a StreamWriter. Prior to this design
// the writer had a single `closed bool` that was flipped to true before the
// terminal encode+publish; a codec error or transport failure would then leave
// the writer wedged in `closed` with no terminal frame on the wire, and the
// peer would only notice via the idle timeout. The state machine below makes
// the "publish in progress" and "publish failed" cases first-class so the
// runtime can react (fallback error, observability, mutual exclusion of
// concurrent terminal calls) instead of silently swallowing the error.
type writerState int

const (
	writerOpen    writerState = iota // Started/Delta permitted; no terminal frame yet
	writerClosing                    // Terminal frame reserved a seq; publish in flight
	writerClosed                     // Terminal frame acknowledged by transport
	writerFailed                     // Terminal frame encode/publish failed; recoverable via forceError
)

type streamWriter struct {
	bus *defaultBus
	ctx context.Context
	req *event.Envelope

	mu      sync.Mutex
	seq     uint64
	started bool
	state   writerState
	// lastErr captures the last codec/publish failure so handleStream can
	// surface it when the handler returned nil but the writer is stuck.
	// Never read while the writer is `open`.
	lastErrVal error
	// reservedSeq holds the Seq that the failed terminal frame consumed.
	// When the writer transitions to `failed`, the seq that was allocated
	// for the doomed terminal frame has never reached the wire — so the
	// client's reorder buffer never saw it. A fallback Error() from FAILED
	// must reuse this seq (not consume a new one), otherwise the client
	// gets a permanent seq hole and idle-times out despite receiving a
	// terminal frame. `reservedValid` distinguishes "no reserved seq" from
	// "reserved seq is 0".
	reservedSeq   uint64
	reservedValid bool
}

func (w *streamWriter) Started(meta any) error {
	w.mu.Lock()
	if err := w.checkOpenLocked(); err != nil {
		w.mu.Unlock()
		return err
	}
	if w.started {
		w.mu.Unlock()
		return errors.New("stream: started already emitted")
	}
	w.started = true
	seq := w.nextSeqLocked()
	w.mu.Unlock()

	env := newReplyShell(w.bus.opts.AgentID, w.req, event.ResponseStarted)
	env.Seq = seq
	if meta != nil {
		data, err := w.bus.opts.Codec.EncodePayload(meta)
		if err != nil {
			return err
		}
		env.Payload = data
	}
	return w.publish(env)
}

func (w *streamWriter) Delta(chunk any) error {
	w.mu.Lock()
	if err := w.checkOpenLocked(); err != nil {
		w.mu.Unlock()
		return err
	}
	seq := w.nextSeqLocked()
	w.mu.Unlock()

	env := newReplyShell(w.bus.opts.AgentID, w.req, event.ResponseDelta)
	env.Seq = seq
	if chunk != nil {
		data, err := w.bus.opts.Codec.EncodePayload(chunk)
		if err != nil {
			return err
		}
		env.Payload = data
	}
	return w.publish(env)
}

// Final publishes a terminal ResponseFinal frame. Unlike the pre-P0#3
// implementation, the writer only transitions to `closed` after publish
// succeeds. If payload encoding or publish fails the writer transitions to
// `failed` and the error is returned to the caller so retry/fallback logic
// (e.g. handleStream's auto-terminal) can observe it. Callers may recover a
// failed writer by calling Error(...), which the state machine still permits.
func (w *streamWriter) Final(result any) error {
	w.mu.Lock()
	if err := w.checkOpenLocked(); err != nil {
		w.mu.Unlock()
		return err
	}
	w.state = writerClosing
	seq := w.nextSeqLocked()
	w.mu.Unlock()

	env := newReplyShell(w.bus.opts.AgentID, w.req, event.ResponseFinal)
	env.Seq = seq
	env.IsFinal = true
	if result != nil {
		data, err := w.bus.opts.Codec.EncodePayload(result)
		if err != nil {
			w.markFailedReserving(err, seq)
			return err
		}
		env.Payload = data
	}
	if err := w.publish(env); err != nil {
		w.markFailedReserving(err, seq)
		return err
	}
	w.markClosed()
	return nil
}

// Error publishes a terminal ResponseError frame. Like Final, the writer
// only reaches `closed` when publish succeeds; a publish failure leaves the
// writer in `failed` and returns the error. `Error(nil)` is a caller bug —
// there is no meaningful payload to encode — and we reject it explicitly
// rather than dereferencing a nil error further down.
func (w *streamWriter) Error(srcErr error) error {
	if srcErr == nil {
		return errors.New("stream: Error requires non-nil error")
	}

	w.mu.Lock()
	// Error is permitted from both `open` (handler decided to fail) and
	// `failed` (a prior Final couldn't reach the wire — try to publish
	// an error frame instead so the peer still sees a terminal).
	if w.state != writerOpen && w.state != writerFailed {
		st := w.state
		w.mu.Unlock()
		return terminalStateErrorFor(st)
	}
	// Reuse the reserved seq from a prior failed terminal so the client's
	// reorder buffer sees a contiguous sequence. If we allocated a fresh
	// seq here, the failed terminal's reserved-but-unpublished seq would
	// stay missing forever and the client would idle-time-out even though
	// this fallback Error frame reached it.
	var seq uint64
	if w.state == writerFailed && w.reservedValid {
		seq = w.reservedSeq
		w.reservedValid = false
	} else {
		seq = w.nextSeqLocked()
	}
	w.state = writerClosing
	w.mu.Unlock()

	env := newReplyShell(w.bus.opts.AgentID, w.req, event.ResponseError)
	env.Seq = seq
	env.IsFinal = true
	payload := event.ErrorPayload{
		Code:    event.CodeAgentUnavailable,
		Message: srcErr.Error(),
	}
	// Encode into a local var first — on encode failure we still transition
	// to `failed` so callers see a consistent state.
	data, err := w.bus.opts.Codec.EncodePayload(payload)
	if err != nil {
		w.markFailedReserving(err, seq)
		return err
	}
	env.Payload = data
	if err := w.publish(env); err != nil {
		w.markFailedReserving(err, seq)
		return err
	}
	w.markClosed()
	return nil
}

// forceError is the internal fallback path used by handleStream when the
// user's handler returned but the writer is in `failed`. It uses the same
// Error() code path so that concurrent guards and observability are shared.
func (w *streamWriter) forceError(src error) error {
	return w.Error(src)
}

func (w *streamWriter) checkOpenLocked() error {
	switch w.state {
	case writerOpen:
		return nil
	case writerClosing:
		return errors.New("stream: terminal frame in flight")
	case writerClosed:
		return errors.New("stream: already closed")
	case writerFailed:
		return errors.New("stream: writer failed; only Error is permitted")
	default:
		return errors.New("stream: unknown writer state")
	}
}

func terminalStateErrorFor(s writerState) error {
	switch s {
	case writerClosing:
		return errors.New("stream: terminal frame in flight")
	case writerClosed:
		return errors.New("stream: already closed")
	default:
		return errors.New("stream: writer in unexpected state")
	}
}

func (w *streamWriter) markClosed() {
	w.mu.Lock()
	w.state = writerClosed
	w.lastErrVal = nil
	// A successful terminal has consumed its seq; clear any prior
	// reservation so a defensive future call doesn't accidentally reuse it.
	w.reservedValid = false
	w.mu.Unlock()
}

func (w *streamWriter) markFailed(err error) {
	w.mu.Lock()
	w.state = writerFailed
	w.lastErrVal = err
	// No reserved seq — this path is for markers that don't own a seq
	// (currently unused after markFailedReserving was introduced, kept for
	// future callers that may want to fail without a seq context).
	w.reservedValid = false
	w.mu.Unlock()
}

// markFailedReserving is the failure path used by Final/Error: it records
// the seq that the failed terminal reserved so a later fallback Error can
// reuse it and avoid a seq hole in the client's reorder buffer.
func (w *streamWriter) markFailedReserving(err error, seq uint64) {
	w.mu.Lock()
	w.state = writerFailed
	w.lastErrVal = err
	w.reservedSeq = seq
	w.reservedValid = true
	w.mu.Unlock()
}

func (w *streamWriter) stateSnapshot() writerState {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.state
}

func (w *streamWriter) lastErr() error {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.lastErrVal
}

// isClosed retains its historical name for callers outside stream.go, but
// now means "reached a terminal state" (either `closed` or `failed`) — both
// mean the writer will not accept further Started/Delta frames.
func (w *streamWriter) isClosed() bool {
	s := w.stateSnapshot()
	return s == writerClosed || s == writerFailed
}

func (w *streamWriter) nextSeqLocked() uint64 {
	s := w.seq
	w.seq++
	return s
}

func (w *streamWriter) publish(env *event.Envelope) error {
	data, err := w.bus.opts.Codec.EncodeEnvelope(env)
	if err != nil {
		return err
	}
	return w.bus.opts.Transport.Publish(w.ctx, &transport.RawMessage{
		Subject: w.req.ReplyTo,
		Data:    data,
	})
}
